d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java

```diff
@@ -690,24 +741,131 @@ void onRemoval()
+        watcher.onChanged(entry.getKey(), entry.getValue());
+        _log.debug("Notifying watcher of current data for resource {} of type {}: {}",
+            entry.getKey(), _type, entry.getValue());
+      }
+    }
+
+    private void onData(String resourceName, ResourceUpdate data)
+    {
+      if (Objects.equals(_data.get(resourceName), data))
+      {
+        _log.debug("Received resource update data equal to the current data. Will not perform the update.");
+        return;
+      }
+      // null value guard to avoid overwriting the property with null
+      if (data != null && data.isValid())
+      {
+        _data.put(resourceName, data);
+        for (WildcardResourceWatcher watcher : _watchers)
+        {
+          watcher.onChanged(resourceName, data);
+        }
+      }
+      else
+      {
+        if (_type == ResourceType.D2_URI_MAP || _type == ResourceType.D2_URI)
+        {
+          RATE_LIMITED_LOGGER.warn("Received invalid data for {} {}, data: {}", _type, resourceName, data);
+        }
+        else
+        {
+          _log.warn("Received invalid data for {} {}, data: {}", _type, resourceName, data);
+        }
+      }
+    }
+
+    public ResourceType getType()
+    {
+      return _type;
+    }
+
+    private void onError(Status error)
+    {
+      for (WildcardResourceWatcher watcher : _watchers)
+      {
+        watcher.onError(error);
+      }
+    }
+
+    private void onReconnect()
+    {
+      for (WildcardResourceWatcher watcher : _watchers)
+      {
+        watcher.onReconnect();
+      }
+    }
+
+    @VisibleForTesting
+    void onRemoval(String resourceName)
+    {
+      _data.remove(resourceName);
+      for (WildcardResourceWatcher watcher : _watchers)
+      {
+        watcher.onRemoval(resourceName);
+      }
+    }
+  }
+
+  final class RpcRetryTask implements Runnable
+  {
     @Override
-    public void run() {
+    public void run()
+    {
       startRpcStreamLocal();
-      for (ResourceType type : ResourceType.values()) {
-        Collection<String> resources = getResourceSubscriberMap(type).keySet();
-        if (resources.isEmpty())
+      for (ResourceType type : ResourceType.values())
+      {
+        Set<String> resources = new HashSet<>(getResourceSubscriberMap(type).keySet());
+        if (resources.isEmpty() && getWildcardResourceSubscriber(type) == null)
         {
           continue;
         }
+        ResourceType rewrittenType;
         if (_subscribeToUriGlobCollection && type == ResourceType.D2_URI_MAP)
         {
           resources = resources.stream()
               .map(GlobCollectionUtils::globCollectionUrlForClusterResource)
-              .collect(Collectors.toSet());
-          type = ResourceType.D2_URI;
+              .collect(Collectors.toCollection(HashSet::new));
+          rewrittenType = ResourceType.D2_URI;
+        }
+        else
+        {
+          rewrittenType = type;
+        }
+        if (getWildcardResourceSubscriber(type) != null)
```

Comment:
It took me a while to understand what's going on here -- looks like we introduce `rewrittenType` because we need the original `type` for the call to `getWildcardResourceSubscriber(type)`, right?

I think one option could be to put this `if` above `if (_subscribeToUriGlobCollection && type == ResourceType.D2_URI_MAP)`, so it gets the original type before reassigning it, that might be cleaner. Either way, it might also be worth adding a comment about the need to look up by the original type before transforming the `D2UriMap` request into a `D2URI` glob request

Reply:
Actually, move before `if (_subscribeToUriGlobCollection && type == ResourceType.D2_URI_MAP)` makes no sense, as the resources is used to create new list 

```java
 resources = resources.stream()
              .map(GlobCollectionUtils::globCollectionUrlForClusterResource)
              .collect(Collectors.toCollection(HashSet::new));
          rewrittenType = ResourceType.D2_URI;
```
So I prefer add comment there.